### PR TITLE
Fixed: PVR functionality NPO

### LINKED ISSUE #1753: PVR stopped working due to API change
### Describe the bug
No channels are found by IPTV Manager, because https://start-api.npo.nl has been deprecated. Refreshing channels results in Error.

## Possible Fix
Change to https://npo.nl/start/api/domain/guide-channel API, just like in recent items.

Note: I will fix it myself. Issue for reference purposes.

### PR DESCRIPTION
**Pull request by landgenoot**
### Functional description
Fixes the PVR/IPTV Manager integration of the NPO channel. Fixes #1753

image
image

### Technical description
This PR removes the use of the **start-api.npo.nl**, and adopts **npo.nl/start/api/domain/guide-channels** in _create_iptv_streams_ and _create_iptv_epg_ of the nos2010 channel. These functions are called by IPTV Manager in order to integrate the live TV channels with the Kodi PVR.

### Discussion 

- The new NPO API does not use channel icons. They switched to SVG files that are embedded in code. Therefore I referenced to the corresponding icons on Wikipedia, to prevent copyright issues. 
- The _guide-channels_-API endpoint does not offer extra details, such as synopsis or genres. The only way seems to be to call _program-details_ for every program. However, this would result in a lot of extra traffic to the NPO. Any ideas on how to show the synopsis?
- _create_api_epg_item()_ returns None if the starttime of the program has not yet passed. IPTV Manager updates the EPG once every 24 hours. Therefore I would like to also create epg items for programs in the future. What is the reason behind this condition?
### Zip
https://codeload.github.com/landgenoot/plugin.video.retrospect/zip/refs/heads/bug-1753

### CODE REVIEW

**Comment by Heleen-Heleen** [comment]
echt geweldig, Enorm bedankt!

File: channels/channel.nos/nos2010/chn_nos2010.py [hunk 1]
```diff
@@ -1165,45 +1165,44 @@ def update_video_item_live(self, item: MediaItem) -> MediaItem:
...
-            livestream = JsonHelper.get_from(stations, "channel", "liveStream")
-            item = MediaItem(JsonHelper.get_from(livestream, "title"),
-                             JsonHelper.get_from(livestream, "shareUrl"),
-                             media_type=mediatype.VIDEO)
-            item.isLive = True
-            item.isGeoLocked = True
+        # NPO start switched to SVG logos embedded in code. So, get them from Wikipedia instead, which should be stable-ish.
+        # 800px width recommended https://kodi.wiki/view/Live_TV_Artwork
+        logo_sources = {
+            "NPO1": "https://upload.wikimedia.org/wikipedia/commons/thumb/0/02/NPO_1_logo_2014.svg/800px-NPO_1_logo_2014.svg.png",
```
Comment (basrieter) [inline channels/channel.nos/nos2010/chn_nos2010.py:1184]: Please use the built-in images: TextureHandler.instance().get_texture_uri(self, "<image name>")

**Review by basrieter — COMMENTED** [review]
What idea did have for the "future" items? That I now filter out.

File: channels/channel.nos/nos2010/chn_nos2010.py [hunk 2]
```diff
@@ -1165,45 +1165,44 @@ def update_video_item_live(self, item: MediaItem) -> MediaItem:
...
+            "NPO2": "https://upload.wikimedia.org/wikipedia/commons/thumb/f/f3/NPO_2_logo_2014.svg/800px-NPO_2_logo_2014.svg.png",
+            "NPO3": "https://upload.wikimedia.org/wikipedia/commons/thumb/c/c9/NPO_3_logo_2014.svg/800px-NPO_3_logo_2014.svg.png",
+            "NPO1 Extra": "https://upload.wikimedia.org/wikipedia/commons/thumb/1/1e/NPO_1_Extra_logo_2018.svg/800px-NPO_1_Extra_logo_2018.svg.png",
+            "NPO2 Extra": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d5/NPO_2_Extra_logo_2018.svg/800px-NPO_2_Extra_logo_2018.svg.png",
```
Comment (basrieter) [inline channels/channel.nos/nos2010/chn_nos2010.py:1188]: My experience is that none of these work, except NPO1/2/3

**Comment by landgenoot** [comment]
> What idea did have for the "future" items? That I now filter out.

If I just comment out this condition on line 736-737, it works fine for me. Because programs in the future that are unavailable don't have a _program_ property. Therefore None is already returned (line 717-729)

https://github.com/retrospect-addon/plugin.video.retrospect/blob/330f7fc88510aade7adc14256fd74165522bc7f7/channels/channel.nos/nos2010/chn_nos2010.py#L736-L737
https://github.com/retrospect-addon/plugin.video.retrospect/blob/330f7fc88510aade7adc14256fd74165522bc7f7/channels/channel.nos/nos2010/chn_nos2010.py#L717-L729

File: channels/channel.nos/nos2010/chn_nos2010.py (hunk 1 above)
Comment (landgenoot) [reply channels/channel.nos/nos2010/chn_nos2010.py]: Sure!

**Comment by basrieter** [comment]
> > What idea did have for the "future" items? That I now filter out.
> 
> If I just comment out this condition on line 736-737, it works fine for me. Because programs in the future that are unavailable don't have a _program_ property. Therefore None is already returned (line 717-729)
> 
> https://github.com/retrospect-addon/plugin.video.retrospect/blob/330f7fc88510aade7adc14256fd74165522bc7f7/channels/channel.nos/nos2010/chn_nos2010.py#L736-L737
> 
> 
> https://github.com/retrospect-addon/plugin.video.retrospect/blob/330f7fc88510aade7adc14256fd74165522bc7f7/channels/channel.nos/nos2010/chn_nos2010.py#L717-L729
> 
> Removing this condition (736-737) will also add the future-but-already-available programs to Recent in Retrosepct.

You are right! Could you comment those 2 lines and a bit a of comment to it, so it is clear why that check is no longer needed?

**Comment by basrieter** [comment]
If you rebase on master the tests will run.

File: channels/channel.nos/nos2010/chn_nos2010.py [hunk 3]
```diff
@@ -1165,45 +1168,44 @@ def update_video_item_live(self, item: MediaItem) -> MediaItem:
...
-            livestream = JsonHelper.get_from(stations, "channel", "liveStream")
-            item = MediaItem(JsonHelper.get_from(livestream, "title"),
-                             JsonHelper.get_from(livestream, "shareUrl"),
-                             media_type=mediatype.VIDEO)
-            item.isLive = True
-            item.isGeoLocked = True
+        # NPO start switched to SVG logos embedded in code. So, get them from Wikipedia instead, which should be stable-ish.
+        # 800px width recommended https://kodi.wiki/view/Live_TV_Artwork
+        logo_sources = {
+            "NPO1": TextureHandler.instance().get_texture_uri(self, "https://upload.wikimedia.org/wikipedia/commons/thumb/0/02/NPO_1_logo_2014.svg/800px-NPO_1_logo_2014.svg.png"),
```
[the remark this answers is not in the document]
Comment (basrieter) [reply channels/channel.nos/nos2010/chn_nos2010.py:1187]: If more are needed, you can add them. However, as a rule of thumb I make icons no longer than 1024x1024 and they should be either JPG or PNG. If PNG is used, please use https://pnggauntlet.com/ to optimize the PNG's.

[the remark this answers is not in the document]
Comment (landgenoot) [reply channels/channel.nos/nos2010/chn_nos2010.py:1187]: Now I understand what you mean. 
NB: These NPO 1-3 logo's are old and [were used 2003-2014](https://logos.fandom.com/wiki/NPO_1). 
Would you mind if I update them?
> The others won't play anything anyway.

Do you mean the streams? For me the extra channels are playing fine.

[the remark this answers is not in the document]
Comment (basrieter) [reply channels/channel.nos/nos2010/chn_nos2010.py:1187]: The live stream do, but not the individual programs. 

Updating the logos is fine, just keep in mind the requirements I mentioned (and keep the backgrounds transparant). Feel free to give them proper names.

[the remark this answers is not in the document]
Comment (landgenoot) [reply channels/channel.nos/nos2010/chn_nos2010.py]: Done.
Interesting to see how effective the PNG optimization is, even after setting the compression level in GIMP to max.

[the remark this answers is not in the document]
Comment (basrieter) [reply channels/channel.nos/nos2010/chn_nos2010.py]: Nice! Ik ga morgen even rustig kijken en zorgen dat het er in komt. Zit met Videoland nog even met wat dingen. En ik zal ook de Retrospect resource add-on (met de images) moeten bijwerken. Dat gaat allemaal geautomatiseerd, maar toch.

[the remark this answers is not in the document]
Comment (basrieter) [reply channels/channel.nos/nos2010/chn_nos2010.py]: Wat ik wel nog zie @landgenoot  is dat je de unit tests even moet fixen. Een aantal texture tests verwezen nog naar de oude PNG files.

**Review by basrieter — APPROVED** [review]
Thanks for all your effort.

File: channels/channel.nos/nos2010/chn_nos2010.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (basrieter) [reply channels/channel.nos/nos2010/chn_nos2010.py]: Klinkt als iets dat wel leuk is! Wel graag in een nieuwe PR.

**Comment by Heleen-Heleen** [comment]
Helemaal top, Heel veel dank mannen!

**Pull request merged**